Travis: Tear out python3 testing to get CI working again

geky opened the pull request:
Temporary solution, python3 will be added back when working (see #6200)

Travis doesn't recognize version matrices when inside a matrix include job. Instead, it breaks in a way that isn't reported as an error.

cc @theotherjimmy, @0xc0170, @deepikabhavnani, @cmonr 
intended for patch
introduced in #5848
related https://github.com/travis-ci/travis-ci/issues/9275

cmonr commented:
/morph build

geky commented:
@cmonr, sorry it's not ready yet. Now, travis fails successfully, but we need to fix the tools so it does not.

cmonr commented:
@geky Oh, my mistake. Was under the impression that the quick fix was ready. In that case, LMK when it is.

geky commented:
Bugs introduced while broken:
- [ ]  tools/test/config/config_test.py:92:
   ```
   assert "Additional properties are not allowed ('unknown_key' was unexpected)" in "Additional properties are not allowed (u'unknown_key' was unexpected)"
   ```
   bisected to: e9fddb6eee25ebbc385bb724fad7918402142e43
   pr: https://github.com/ARMmbed/mbed-os/pull/6017

- [ ]   tools/toolchains/arm.py:307:
   ```
   KeyError: 'build_dir'
   ```
   bisected to: b21e93ba32df17a283c96f32e595ff69f62a456d
   pr: https://github.com/ARMmbed/mbed-os/pull/6096

theotherjimmy commented:
#6096 did not have to add the offending line; it added nothing.

theotherjimmy commented:
The first error is a unicode problem XD notice the `u'...'`

deepikabhavnani commented:
> #6096 did not have to add the offending line; it added nothing.

@theotherjimmy  - offending line was a suggestion from your side to get the path of build directory, which is used few steps below. 
`secure_file = join(build_dir, "cmse_lib.o")`

Will appreciate to have proper fix from your side, to append the build directory path for cmse_lib.o creation.

geky commented:
I added some fixes/workarounds, @theotherjimmy, @deepikabhavnani, if you could review and let me know if you're happy with these for now.

> The first error is a unicode problem XD notice the u'...'

Haha, it's not even a _real_ unicode error. It's just that now the config will show that "u" when it has errors, and the tests are checking the error strings.

> offending line was a suggestion from your side to get the path of build directory, which is used few steps below.

@deepikabhavnani, I moved the `kwargs['build_dir']` into the condition for "secure". This lets the test pass for now, let me know if you're happy with this as a workaround so we can get tool testing back into mbed OS.

theotherjimmy commented:
@geky That workaround works, but I would prefer that the offending test just pass in a dummy `build_dir` parameter. @deepikabhavnani My bad, I thought that line was in the ARM_STD constructor. It makes sense where it is.

deepikabhavnani commented:
> I moved the kwargs['build_dir'] into the condition for "secure". This lets the test pass for now, let me know if you're happy with this as a workaround so we can get tool testing back into mbed OS.

@theotherjimmy - Can you please confirm build_dir will not be `none` for Cortex-M23/M33 devices. 

Toolchain init is different for ARM / ARMC6, will this have any impact?
https://github.com/geky/mbed/blob/962be07a31586f6cc8d8e2ae8dddb7b05b7e4554/tools/toolchains/arm.py#L49
https://github.com/geky/mbed/blob/962be07a31586f6cc8d8e2ae8dddb7b05b7e4554/tools/toolchains/arm.py#L307

theotherjimmy commented:
@deepikabhavnani Build dir being None is an error. We need a place to put object files.

theotherjimmy commented:
@geky Feel free to delete the `test_instantiation` test. It adds no value anymore.

theotherjimmy commented:
It added value at one point, but no longer.

geky commented:
> @geky Feel free to delete the test_instantiation test. It adds no value anymore.

Cool beans, it's gone now.

@cmonr Should be good to go now. Though no rush, travis is severly backlogged because of mbed 2 builds.

geky commented:
@cmonr, sorry for the slip up. Should be good to go now (heavy handed solution).

I moved the build dir back into the condition. It looks like none of the other tests pass in build_dir either, but the root of this problem can be fixed in another pr.

cmonr commented:
/morph build

adbridge commented:
Reverting a change that only went in for 5.8 so should have also targeted that release!
